primitives/npos-elections/solution-type/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Proc macro for a npos solution type.
+
+use proc_macro::TokenStream;
+use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
+use proc_macro_crate::{crate_name, FoundCrate};
+use quote::quote;
+use syn::parse::{Parse, ParseStream, Result};
+
+mod codec;
+mod from_assignment_helpers;
+mod index_assignment;
+mod multi_page;
+mod single_page;
+
+/// Get the name of a filed based on voter count.
+pub(crate) fn vote_filed(n: usize) -> Ident {
+	quote::format_ident!("votes{}", n)
+}
+
+/// Get the name of a filed based on pae count.
+pub(crate) fn page_field(page: u8) -> Ident {
+	quote::format_ident!("page{}", page)
+}
+
+/// Generate a `syn::Error`.
+pub(crate) fn syn_err(message: &'static str) -> syn::Error {
+	syn::Error::new(Span::call_site(), message)
+}
+
+/// Generates a struct to store the election result in a small/compact way. This can encode a
+/// structure which is the equivalent of a `sp_npos_elections::Assignment<_>`.
+///
+/// The following data types can be configured by the macro.
+///
+/// - The identifier of the voter. This can be any type that supports `parity-scale-codec`'s compact
+///   encoding.
+/// - The identifier of the target. This can be any type that supports `parity-scale-codec`'s
+///   compact encoding.
+/// - The accuracy of the ratios. This must be one of the `PerThing` types defined in
+///   `sp-arithmetic`.
+///
+/// Moreover, the maximum number of edges per voter (distribution per assignment) also need to be
+/// specified. Attempting to convert from/to an assignment with more distributions will fail.
+///
+/// For example, the following generates a public struct with name `TestSolution` with `u16` voter
+/// type, `u8` target type and `Perbill` accuracy with maximum of 4 edges per voter.
+///
+/// ```
+/// # use sp_npos_elections_solution_type::generate_solution_type;
+/// # use sp_arithmetic::per_things::Perbill;
+/// generate_solution_type!(pub struct TestSolution::<
+///     VoterIndex = u16,
+///     TargetIndex = u8,
+///     Accuracy = Perbill,
+/// >(4));
+/// ```
+///
+/// The output of this macro will roughly look like:
+///
+/// ```ignore
+/// struct TestSolution {
+///     voters1: vec![(V, T), ..]
+///     voters2: vec![V, P, [(T1, P1), (T2, P2)], T_last]
+///     voters3: <stripped>
```

Comment:
also is the example wrong ? here is I think the correct fix
```suggestion
///     voters1: vec![(u16 /* voter */, u8 /* target */)]
///     voters2: vec![(u16 /* voter */, (u8 /* first target*/, Perbill /* proportion for first target */ ), u8 /* last target */)]
///     voters3: vec![(u16 /* voter */,  [(u8 /* first target*/, Perbill /* proportion for first target */ ), (u8 /* second target */, Perbill /* proportion for second target*/)], u8 /* last target */)],
```

Reply:
I was missing a tuple, fixed it.